core/internal/runfiles/uploader.go

```diff
@@ -119,3 +156,105 @@ func (u *uploader) lockForOperation(method string) bool {
 
 	return true
 }
+
+// Uploads the given files unless we are offline.
+//
+// This increments `uploadWG` and returns immediately. The wait group is
+// signalled as file uploads finish.
+func (u *uploader) upload(relativePaths []string) {
+	if u.settings.IsOffline() {
+		return
+	}
+
+	relativePaths = u.filterNonExistingAndWarn(relativePaths)
+	u.uploadWG.Add(len(relativePaths))
+
+	go func() {
+		createRunFilesResponse, err := gql.CreateRunFiles(
+			context.Background(),
+			u.graphQL,
+			u.settings.GetEntity(),
+			u.settings.GetProject(),
+			u.settings.GetRunID(),
+			relativePaths,
```

Comment:
nice, i can see a future optimization where we can batch some of the file upload request and save on this...

Reply:
yep! we do this optimization for "UploadRemaining" right now as well 😄